lib/services.cf

```diff
@@ -1098,3 +1104,74 @@ bundle agent classic_services(service,state)
       "DEBUG $(this.bundle): The baseinit is NOT provided, using default"
       if => not(isvariable("baseinit[$(service)]"));
 }
+
+body service_method aix_service_method
+{
+  service_bundle => aix_services("$(this.promiser)","$(this.service_policy)");
+}
+
+# example of querying state of a service on AIX
+#
+# bash-5.1# /usr/bin/lssrc -s sendmail
+# Subsystem         Group            PID          Status
+#  sendmail         mail             5308762      active
+#
+# according to https://docs.cfengine.com/docs/3.24/reference-promise-types-services.html#service_policy
+# state can be one of start, stop, enable, disable, restart and reload.
+# disable/enable might be available for services in /etc/inetd.conf
+# e.g. https://www.ibm.com/support/pages/ibm-aix-how-disable-rsh-and-rlogin-services
+# /usr/bin/lssrc -t login, comment out lines in /etc/inetd.conf
+#
+# Note: This service method bundle does NOT handle inetd services like rsh/rlogin
+# only subsystems aka those services listed with /usr/bin/lssrc -a
+# https://www.ibm.com/docs/en/aix/7.3?topic=daemons-subsystems-subservers
+#
+# Also note this from the lssrc man page:
+#        The lssrc command output can sometimes show two entries for a particular daemon. One instance will be active and another instance will be
+#       inoperative. This can happen if the subsystem is modified (using the mkssys command or chssys command) without stopping the subsystem. The
+#       original subsystem will remain active and the modified instance will be inoperative until the subsystem is stopped and started again.
+#
+# Additional output may appear prefixed with Q: for example if it takes some time to change the service state:
+#
+#   notice: Q: "...in/stopsrc -s s": 0513-056 Timeout waiting for command response. If you specified a foreign host,
+# Q: "...in/stopsrc -s s": see the /etc/inittab file on the foreign host to verify that the SRC daemon
+# Q: "...in/stopsrc -s s": (srcmstr) was started with the -r flag to accept remote requests.
+# Q: "...in/stopsrc -s s": 0513-059 The sendmail Subsystem has been started. Subsystem PID is 13042068.
+#
+# Another failure case. Likely on next agent run (default of 5 minutes) the service will have stopped or we will try again.
+#    error: Finished command related to promiser '/usr/bin/stopsrc -s sendmail' -- an error occurred, returned 1
+#   notice: Q: "...in/stopsrc -s s": 0513-056 Timeout waiting for command response. If you specified a foreign host,
+# Q: "...in/stopsrc -s s": see the /etc/inittab file on the foreign host to verify that the SRC daemon
+# Q: "...in/stopsrc -s s": (srcmstr) was started with the -r flag to accept remote requests.
+bundle agent aix_services(service, desired_state)
+{
+  vars:
+    # current state can be: active, inoperative, stopping
+    "current_state" string => execresult("/usr/bin/lssrc -s $(service) | tail -1 | awk '{print $NF}'", useshell);
```

Comment:
I have some small amount of worry about cached result here, it *might* benefit from action => immediate but I think it should be ok without it since $(service) comes from the parameter and that won't be set until real 3 pass actuation and not during pre-eval.

Reply:
Ok. :) Yeah, ideally folks shouldn't have to worry about it right?

What would be the problem about "cached result"?
I know that the commands in execresult are run a few times and cached in pre-eval and then in the real passes right?
Seems OK to me regardless as things will "work out"/stabilize eventually.

Reply:
The problem with the cached result is that the state from pre-eval could be used instead of the state from when this is actuated.

Again, I don't think it applies here because $(service) would not be defined during pre-eval since it's only available as a parameter which will only happen during the normal 3 pass evaluation.

However, if this bundle happens to be called for the same service multiple times within the policy, then the cache is likely to apply which could cause some issue if you for example do:

start (but say it was already running), stop, start

In that second start the cached value from the first start would be used and with the knowledge it was running it may not try to start again.

Reply:
Yeah. I suppose it would be good to include a warning of some sort if you try to change the state of a service that has already been managed in the current agent run... kind of like a global class that signified e.g. "sshd_service_managed" and if you re-entered the aix_services() bundle and that class was set then log a warning that you shouldn't try to enforce more than one state on a service in a policy run.

Reply:
I made a ticket: https://northerntech.atlassian.net/browse/CFE-4448